### PR TITLE
Update object detection guide

### PR DESCRIPTION
**Pull request by qubvel**
# What does this PR do?

Updates object detection guide with better model training and evaluation routines.
Inspired by https://github.com/huggingface/transformers/pull/30422

Relates to 
https://github.com/huggingface/transformers/issues/29964
https://github.com/huggingface/transformers/issues/30557

## Before submitting
- [x] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by qubvel** [comment]
@amyeroberts @NielsRogge please review the updated object detection guide, if everything is fine I will update the notebook example also.
[Docs link](https://moon-ci-docs.huggingface.co/docs/transformers/pr_30683/en/tasks/object_detection)

File: docs/source/en/tasks/object_detection.md [hunk 1]
```diff
@@ -394,211 +1441,118 @@ Hugging Face Hub. Upon training completion, push the final model to the Hub as w
...
->>> model = AutoModelForObjectDetection.from_pretrained("devonho/detr-resnet-50_finetuned_cppe5")
+>>> np_preprocessed_image = resize_and_pad(image=np.array(image))["image"]
 
 >>> with torch.no_grad():
-...     inputs = image_processor(images=image, return_tensors="pt")
-...     outputs = model(**inputs)
-...     target_sizes = torch.tensor([image.size[::-1]])
-...     results = image_processor.post_process_object_detection(outputs, threshold=0.5, target_sizes=target_sizes)[0]
+...     inputs = image_processor(images=[np_preprocessed_image], return_tensors="pt")
+...     outputs = model(inputs["pixel_values"].cuda())
```
Comment (amyeroberts) [inline docs/source/en/tasks/object_detection.md:1523]: nit - here we're assuming there's a cuda device available and the model has been loaded onto it. I'd make it mode explicit by defining a `device` var which we then do `.to(device)` for both the inputs and the model. This way, it's a simple one-line change for someone to run on e.g. cpu or xpu.

**Review by amyeroberts — APPROVED** [review]
Beautiful - thanks for working on this and updating!

File: docs/source/en/tasks/object_detection.md [hunk 2]
```diff
@@ -309,18 +376,137 @@ Pad images (which are now `pixel_values`) to the largest image in a batch, and c
...
+...     data["labels"] = [x["labels"] for x in batch]
+...     if "pixel_mask" in batch[0]:
+...         data["pixel_mask"] = torch.stack([x["pixel_mask"] for x in batch])
+...     return data
+
+```
+
+## Preparing function to compute mAP
+
+Object detection models are commonly evaluated with a set of <a href="https://cocodataset.org/#detection-eval">COCO-style metrics</a>. We are going to use `torchmetrics` to compute `mAP` and `mAR` metrics and will wrap it to `compute_metrics` function in order to use in [`Trainer`] for evaluation.
```
Comment (NielsRogge) [inline docs/source/en/tasks/object_detection.md:393]:
```suggestion
Object detection models are commonly evaluated with a set of <a href="https://cocodataset.org/#detection-eval">COCO-style metrics</a>. We are going to use `torchmetrics` to compute `mAP` (mean average precision) and `mAR` (mean average recall) metrics and will wrap it to `compute_metrics` function in order to use in [`Trainer`] for evaluation.
```

**Review by NielsRogge — APPROVED** [review]
Thanks a lot for improving this guide!

File: docs/source/en/tasks/object_detection.md [hunk 3]
```diff
@@ -150,27 +171,21 @@ To get an even better understanding of the data, visualize an example in the dat
...
+    <img src="https://i.imgur.com/oVQb9SF.png" alt="CPPE-5 Image Example"/>
 </div>
 
+
 To visualize the bounding boxes with associated labels, you can get the labels from the dataset's metadata, specifically
 the `category` field.
 You'll also want to create dictionaries that map a label id to a label class (`id2label`) and the other way around (`label2id`).
 You can use them later when setting up the model. Including these maps will make your model reusable by others if you share
-it on the Hugging Face Hub. Please note that, the part of above code that draws the bounding boxes assume that it is in `XYWH` (x,y co-ordinates and width and height of the box) format. It might not work for other formats like `(x1, y1, x2, y2)`.  
+it on the Hugging Face Hub. Please note that, the part of above code that draws the bounding boxes assume that it is in `COCO` format `(x_min, y_min, width, height)`. It has to be adjusted to work for other formats like `(x_min, y_min, x_max, y_max)`.
```
[the remark this answers is not in the document]
Comment (qubvel) [reply docs/source/en/tasks/object_detection.md:183]: Thanks for reviewing!

File: docs/source/en/tasks/object_detection.md (hunk 1 above)
Comment (qubvel) [reply docs/source/en/tasks/object_detection.md]: good point!

File: docs/source/en/tasks/object_detection.md (hunk 2 above)
Comment (qubvel) [reply docs/source/en/tasks/object_detection.md]: Thanks for the review!
